### PR TITLE
0.0.1

### PR DESCRIPTION
**Pull request by georgehrke**
🚧 highly work-in-progress 🚧

### CODE REVIEW

File: src/factories/davCollectionPublishableFactory.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import * as NS from "../utility/namespaceUtility";
+
+
+function publishedAndPublicToken(props) {
```
Comment (tcitworld) [inline src/factories/davCollectionPublishableFactory.js]: missing export ?

**Comment by georgehrke** [comment]
```javascript
const client = new cdav.DavClient({rootUrl: 'http://all.local/nextcloud/remote.php/dav'}, xhrProvider);
await client.connect({enableCardDAV: true});
```

**Comment by skjnldsv** [comment]
@georgehrke I should have fixed stuff, but I forgot the fixes we made together, can you take a look?
I get a import error on contacts now :)

**Comment by georgehrke** [comment]
308bf8e4-6d66-4524-80d4-b650bd024f32-2

ea2a8d3e-2f71-4218-ba7f-dca06068b58b

tl;dr:
- you should use `findAllAddressBooks()` instead of `findAll()` to make sure you only get address books and not any other DavCollection thats in the directory.
- known issue: you can't update contacts data on the same object more than once, because we don't store the updated e-tag yet. Hence you will get a precondition-failed on any further update attempt.

**Comment by georgehrke** [comment]
There are two things I'm generally not happy about right now:
- The way our parsers work and how davclient.js parses the data. I think our parsers could be much more efficient if we kicked out davclient.js (We don't use 75% of davclient's functionality anyway) and if we are able to properly use XPATH on everything. But that's just a detail inside the library. The libraries interface to other apps won't change.

- I'm not sure whether we need webpack at all. I managed to write native module based ES6 tests in karma / jasmine. So webpack is no longer required for testing. And this library is mostly used by other apps which use webpack themselves. So I'm considering pointing index to `src/index.js` instead of `dist/dist.js` and just let the apps which use this library do all the webpack stuff. Any opinion? :)

**Comment by skjnldsv** [comment]
> * The way our parsers work and how davclient.js parses the data. I think our parsers could be much more efficient if we kicked out davclient.js (We don't use 75% of davclient's functionality anyway) and if we are able to properly use XPATH on everything. But that's just a detail inside the library. The libraries interface to other apps won't change.

Oh, I thought you wanted to get rid of davclient :)
Yes, we should probably drop it in the future.
This is just a webdav api requests handler, right? Maybe it's time to find a more supported one, then? https://www.npmjs.com/package/webdav

> * I'm not sure whether we need webpack at all. I managed to write native module based ES6 tests in karma / jasmine. So webpack is no longer required for testing. And this library is mostly used by other apps which use webpack themselves. So I'm considering pointing index to `src/index.js` instead of `dist/dist.js` and just let the apps which use this library do all the webpack stuff. Any opinion? :)

Is you app es6 compatible only?
Because I see a lot of `() => {`. Therefore It could be great to enable webpack and babel, because we don't have ie11 support right now :)
My opinion is: it's so much easier to manage webpack files, especially since the v4 is almost a noconfig.

**Comment by skjnldsv** [comment]
Now I fixed it!

```js
import DavClient from 'cdav-library'

function xhrProvider() {
	var headers = {
		'X-Requested-With': 'XMLHttpRequest',
		'requesttoken': OC.requestToken
	}
	var xhr = new XMLHttpRequest()
	var oldOpen = xhr.open

	// override open() method to add headers
	xhr.open = function() {
		var result = oldOpen.apply(this, arguments)
		for (let name in headers) {
			xhr.setRequestHeader(name, headers[name])
		}
		return result
	}
	OC.registerXHRForErrorProcessing(xhr)
	return xhr
}


export default new DavClient({
	rootUrl: OC.linkToRemote('dav')
}, xhrProvider)
```

**Comment by skjnldsv** [comment]
Errors found:
- [ ] when creating an addressbook, the return value is not a valid collection.
- [ ] same goes for vcards
- [ ] more to come? :grin: :

**Comment by skjnldsv** [comment]
- [ ] Url returned on findAll is not the same as on find. We need the last backslash as well.

**Comment by skjnldsv** [comment]
- [ ] We do not use the uid as filename anymore and it is breaking the concurrency check (see https://github.com/nextcloud/server/issues/11387)
- [ ] If the server returns a 400 (for example) on creation, we still return a fetch/find. We should directly fail and not try to find the vObject

File: webpack.common.js [hunk 2]
```diff
@@ -6,6 +6,8 @@ module.exports = {
         path: path.resolve(__dirname, './dist'),
         publicPath: '/',
         filename: 'dist.js',
-        library: 'cdav'
+        library: 'cdav',
+        libraryTarget: 'umd',
+		umdNamedDefine: true
```
Comment (georgehrke) [inline webpack.common.js]: wrong indentation, will fix

Comment (skjnldsv) [reply webpack.common.js]: Oups :see_no_evil:

File: src/models/davObject.js [hunk 3]
```diff
@@ -54,7 +54,7 @@ export class DavObject extends DAVEventListener {
 			_isDirty: false
 		});
 
-		this._exposeProperty('etag', NS.DAV, 'getetag');
+		this._exposeProperty('etag', NS.DAV, 'getetag', true);
```
Comment (georgehrke) [inline src/models/davObject.js]: etag should not be publicly editable, just from inside this class.

Not a big deal though, let's just merge this and i will send a pull-request for this later.

Comment (skjnldsv) [reply src/models/davObject.js]: I added this for us to be able to specifically update an etag in case of errors :)
Otherwise you have to reload the entire app :/
